### PR TITLE
Fix `Function.prototype.call` and tests.

### PR DESCRIPTION
**Pull request by Xotic750**
Ref: https://github.com/es-shims/es5-shim/issues/304

This is not necessarily ready to pull, but I want for the tests to be run.

### CODE REVIEW

**Comment by Xotic750** [comment]
All seems well so far on Node. Is there somewhere to see the automated browser tests (if there are any)?

File: es5-shim.js [hunk 1]
```diff
@@ -6,12 +6,15 @@
 
 // vim: ts=4 sts=4 sw=4 expandtab
 
-// Add semicolon to prevent IIFE from being passed as argument to concatenated code.
-;
+// Added to override editor options
+/*jshint maxparams:false, maxdepth:false, maxstatements:false, maxcomplexity:false,
```
Comment (ljharb) [inline es5-shim.js]: please revert this - we don't use jshint in this repo.

File: es5-shim.js [hunk 2]
```diff
@@ -64,7 +68,6 @@ var to_string = ObjectPrototype.toString;
 
 var hasToStringTag = typeof Symbol === 'function' && typeof Symbol.toStringTag === 'symbol';
 var isCallable; /* inlined from https://npmjs.com/is-callable */ var fnToStr = Function.prototype.toString, tryFunctionObject = function tryFunctionObject(value) { try { fnToStr.call(value); return true; } catch (e) { return false; } }, fnClass = '[object Function]', genClass = '[object GeneratorFunction]'; isCallable = function isCallable(value) { if (typeof value !== 'function') { return false; } if (hasToStringTag) { return tryFunctionObject(value); } var strClass = to_string.call(value); return strClass === fnClass || strClass === genClass; };
-var isRegex; /* inlined from https://npmjs.com/is-regex */ var regexExec = RegExp.prototype.exec, tryRegexExec = function tryRegexExec(value) { try { regexExec.call(value); return true; } catch (e) { return false; } }, regexClass = '[object RegExp]'; isRegex = function isRegex(value) { if (typeof value !== 'object') { return false; } return hasToStringTag ? tryRegexExec(value) : to_string.call(value) === regexClass; };
```
Comment (ljharb) [inline es5-shim.js]: please revert this as well - this is required because `Object#toString` is not reliable in ES6.

File: es5-shim.js [hunk 3]
```diff
@@ -176,6 +179,36 @@ var ES = {
     }
 };
 
+/**
+ * An implementation of `slice`. Primarily for use with `arguments`.
```
Comment (ljharb) [inline es5-shim.js]: we already have `sliceArgs` internally, will this not work?

File: es5-shim.js (hunk 1 above)
Comment (Xotic750) [inline es5-shim.js]: I know, it was just while in my editor, I will remove stuff that has nothing to do with the fix.

File: es5-shim.js (hunk 2 above)
Comment (Xotic750) [inline es5-shim.js]: No problem, just wanted to for my code path for now.

File: es5-shim.js (hunk 3 above)
Comment (Xotic750) [inline es5-shim.js]: Aha, I didn't notice it.

Comment (Xotic750) [inline es5-shim.js]: Can't find `sliceArgs`? Need it instead of `Array#slice` because I believe that in some of the environments that these fixes are for, it doesn't slice arguments.

Comment (ljharb) [inline es5-shim.js]: oh oops, i'm referring to something in the es6-shim :-) https://github.com/paulmillr/es6-shim/blob/master/es6-shim.js#L1252-L1263

File: es5-shim.js [hunk 4]
```diff
@@ -176,6 +177,36 @@ var ES = {
...
+ *
+ * @private
```
Comment (ljharb) [inline es5-shim.js]: also please remove the jsdoc, i don't permit it in any projects i maintain

File: es5-shim.js [hunk 5]
```diff
@@ -176,6 +177,36 @@ var ES = {
...
+ * @param {Array} array The array to slice.
+ * @param {number} [start=0] The start position.
+ * @param {number} [end=array.length] The end position.
+ * @returns {Array} Returns the slice of `array`.
+ */
+var baseSlice = function (array, start, end) {
+  /* jshint eqnull: true */
```
Comment (ljharb) [inline es5-shim.js]: please remove the jshint directive

File: es5-shim.js [hunk 6]
```diff
@@ -329,6 +360,97 @@ var strSplit = call.bind(StringPrototype.split);
 var strIndexOf = call.bind(StringPrototype.indexOf);
 var push = call.bind(array_push);
 
+// ES-5 15.3.4.4
+// http://es5.github.io/#x15.3.4.4
+var hasToStringTagBasicBug = !(toStr() === '[object Undefined]' && toStr(null) === '[object Null]');
```
Comment (ljharb) [inline es5-shim.js]: instead of !(a && b)` could this be `!a || !b`?

File: es5-shim.js [hunk 7]
```diff
@@ -329,6 +360,97 @@ var strSplit = call.bind(StringPrototype.split);
...
+var hasToStringTagArgumentsBug = toStr(function () {
+  return arguments;
+}) !== '[object Arguments]';
```
Comment (ljharb) [inline es5-shim.js]: this isn't a bug so much as a legacy arguments object - which is already checked [here](https://github.com/es-shims/es5-shim/blob/master/es5-shim.js#L932-L945)

File: es5-shim.js (hunk 4 above)
Comment (Xotic750) [inline es5-shim.js]: Ok

File: es5-shim.js (hunk 5 above)
Comment (Xotic750) [inline es5-shim.js]: Will remove all of them

File: es5-shim.js (hunk 6 above)
Comment (Xotic750) [inline es5-shim.js]: Indeed

File: es5-shim.js (hunk 7 above)
Comment (Xotic750) [inline es5-shim.js]: Yep, but it's different in different environments. All ES5 environments give `[object Arguments]`, whether in the spec or not though I'm not sure. Can't use your check as it will cause recursion in the fix.

File: es5-shim.js [hunk 8]
```diff
@@ -329,6 +360,97 @@ var strSplit = call.bind(StringPrototype.split);
...
+var hasToStringTagArrayBug = toStr([]) !== '[object Array]';
```
Comment (ljharb) [inline es5-shim.js]: in what browser is this `true`? The defacto `Array.isArray` shim, which works in every single browser I've ever tested in, is `toStr(foo) === '[object Array]'`

File: es5-shim.js [hunk 9]
```diff
@@ -329,6 +360,97 @@ var strSplit = call.bind(StringPrototype.split);
...
+var hasToStringTagRegExpBug = toStr(/test/) !== '[object RegExp]';
```
Comment (ljharb) [inline es5-shim.js]: similarly, in what browser is this `true`?

File: es5-shim.js [hunk 10]
```diff
@@ -329,6 +360,97 @@ var strSplit = call.bind(StringPrototype.split);
...
+        typeof value.length === 'number' &&
+        value.length >= 0 &&
+        !owns(value, 'callee') &&
+        !owns(value, 'arguments');
+  };
+}
+
+if (hasToStringTagRegExpBug) {
+  var isDuckTypeRegExp = function (value) {
+    /*jshint eqnull:true */
```
Comment (ljharb) [inline es5-shim.js]: remove jshint directive

File: es5-shim.js [hunk 11]
```diff
@@ -329,6 +360,97 @@ var strSplit = call.bind(StringPrototype.split);
...
+    if (value != null) {
+      var type = typeof value;
+      if (type === 'object' || type === 'function') {
+        try {
+          regexExec.call(value);
```
Comment (ljharb) [inline es5-shim.js]: this should be robust against `delete Function.prototype.call`

File: es5-shim.js (hunk 9 above)
Comment (Xotic750) [inline es5-shim.js]: IE6-8 -> hasToStringTagBasicBug
The others I can't remember exactly, but could pick the tests across and run them as I have no way of testing any more.

File: es5-shim.js (hunk 11 above)
Comment (Xotic750) [inline es5-shim.js]: Not sure what you mean?

Comment (ljharb) [inline es5-shim.js]: Any `.call` will fail if `Function.prototype.call` is later broken or nonexistent

Comment (Xotic750) [inline es5-shim.js]: Ok.

File: es5-shim.js [hunk 12]
```diff
@@ -163,7 +164,6 @@ var ES = {
     // http://es5.github.com/#x9.9
     /* replaceable with https://npmjs.com/package/es-abstract ES5.ToObject */
     ToObject: function (o) {
-        /* jshint eqnull: true */
```
Comment (ljharb) [inline es5-shim.js]: hm, i didn't realize this was in master. thanks for removing it!

File: es5-shim.js (hunk 7 above)
Comment (ljharb) [inline es5-shim.js]: Primarily my concern is keeping the chunks of code that are replaceable with existing npm modules together and untouched.

File: es5-shim.js (hunk 9 above)
Comment (ljharb) [inline es5-shim.js]: yes - please don't add any of the checks unless you can include a comment indicating in which browsers it fails.

Comment (ljharb) [inline es5-shim.js]: * specifically, the runtime checks. adding actual spec tests is always OK :-)

File: es5-shim.js (hunk 12 above)
Comment (Xotic750) [inline es5-shim.js]: No problem.

File: es5-shim.js (hunk 11 above)
Comment (Xotic750) [inline es5-shim.js]: Not really sure how this can be achieved. If someone deletes `Function.prototype.call` it seems everything is pretty much fubar, whether es5-shim is loaded or not. Did you have something in mind?

File: es5-shim.js [hunk 13]
```diff
@@ -329,6 +351,94 @@ var strSplit = call.bind(StringPrototype.split);
 var strIndexOf = call.bind(StringPrototype.indexOf);
 var push = call.bind(array_push);
 
+// ES-5 15.3.4.4
+// http://es5.github.io/#x15.3.4.4
+var hasToStringTagBasicBug = toStr() !== '[object Undefined]' || toStr(null) !== '[object Null]';
+var hasToStringTagLegacyArguments = toStr(function () {
+  return arguments;
+}) !== '[object Arguments]';
```
Comment (michaelficarra) [inline es5-shim.js]: You forgot to invoke the function.

Comment (Xotic750) [inline es5-shim.js]: Dope!

File: es5-shim.js [hunk 14]
```diff
@@ -329,6 +351,94 @@ var strSplit = call.bind(StringPrototype.split);
...
+          return false;
+        }
+      }
+    }
+  };
+}
+
+if (hasToStringTagInconsistency) {
+  defineProperties(FunctionPrototype, {
+    call: function () {
```
Comment (michaelficarra) [inline es5-shim.js]: You'll have to do the same thing with `apply`, right? Might want to just have `call` defer to `apply` in that case.

File: es5-shim.js (hunk 11 above)
Comment (ljharb) [inline es5-shim.js]: It's not possible in the sense that the `bind` polyfill relies on it (and has to). However, none of the other functions need to.

In the es5-shim, `.call` is one of the ones I haven't tried too hard to be robust against, but it's still useful to try.

File: es5-shim.js (hunk 14 above)
Comment (Xotic750) [inline es5-shim.js]: Yeah, a similar thing will need to be done with `apply`. Just concentrating on `call` right now, exploring the possibilities, proving that it can be done.

File: es5-shim.js [hunk 15]
```diff
@@ -329,6 +351,94 @@ var strSplit = call.bind(StringPrototype.split);
 var strIndexOf = call.bind(StringPrototype.indexOf);
 var push = call.bind(array_push);
 
+// ES-5 15.3.4.4
+// http://es5.github.io/#x15.3.4.4
+var hasToStringTagBasicBug = toStr() !== '[object Undefined]' || toStr(null) !== '[object Null]';
+var hasToStringTagLegacyArguments = toStr(function () {
+  return arguments;
+}()) !== '[object Arguments]';
+var hasToStringTagArrayBug = toStr([]) !== '[object Array]';
```
Comment (ljharb) [inline es5-shim.js]: just making sure we don't forget to delete this line, unless we can find (and document in a comment) a specific browser that would give `true` here.

File: es5-shim.js [hunk 16]
```diff
@@ -329,6 +351,94 @@ var strSplit = call.bind(StringPrototype.split);
...
+var hasToStringTagRegExpBug = toStr(/test/) !== '[object RegExp]';
```
Comment (ljharb) [inline es5-shim.js]: just making sure we don't forget to delete this line, unless we can find (and document in a comment) a specific browser that would give `true` here.

File: es5-shim.js [hunk 17]
```diff
@@ -329,6 +351,94 @@ var strSplit = call.bind(StringPrototype.split);
 var strIndexOf = call.bind(StringPrototype.indexOf);
 var push = call.bind(array_push);
 
+// ES-5 15.3.4.4
+// http://es5.github.io/#x15.3.4.4
+var hasToStringTagBasicBug = toStr() !== '[object Undefined]' || toStr(null) !== '[object Null]';
+var hasToStringTagLegacyArguments = toStr(function () {
+  return arguments;
+}()) !== '[object Arguments]';
```
Comment (ljharb) [inline es5-shim.js]: in this case, it's already being invoked in an IIFE, so you can just do `toStr(arguments) !== '[object Arguments]'`

File: es5-shim.js [hunk 18]
```diff
@@ -1452,7 +1562,7 @@ defineProperties(NumberPrototype, { toFixed: toFixedShim }, hasToFixedBugs);
 
 var hasToPrecisionUndefinedBug = (function () {
     try {
-        return 1.0.toPrecision(undefined) === '1';
+        return (1.0).toPrecision(undefined) === '1';
```
Comment (ljharb) [inline es5-shim.js]: these parens should not be needed. please revert.

File: es5-shim.js (hunk 17 above)
Comment (Xotic750) [inline es5-shim.js]: Good point! :)

File: es5-shim.js [hunk 19]
```diff
@@ -1654,7 +1764,7 @@ defineProperties(StringPrototype, {
     }
 }, hasTrimWhitespaceBug);
 
-var hasLastIndexBug = StringPrototype.lastIndexOf && 'abcあい'.lastIndexOf('あい', 2) !== -1;
+var hasLastIndexBug = StringPrototype.lastIndexOf && 'abc\u3042\u3044'.lastIndexOf('\u3042\u3044', 2) !== -1;
```
Comment (ljharb) [inline es5-shim.js]: please revert this change as well - the file is already UTF-8.

File: tests/spec/s-function.js [hunk 20]
```diff
@@ -1,7 +1,177 @@
 /* global describe, it, expect, beforeEach */
+var hasStrictMode = (function () {
+  'use strict';
+
+   return !this;
+}());
+
+var global = (function () {
+   return this;
```
Comment (ljharb) [inline tests/spec/s-function.js]: this will fail silently if strict mode has been globally enabled - please ensure that `this` is truthy as well.

File: tests/spec/s-function.js [hunk 21]
```diff
@@ -1,7 +1,177 @@
...
+        it('should not throw a error in each case', function () {
+            expect(function () {
+                noop.call();
+            }).not.toThrow();
+            expect(function () {
+                /*eslint no-useless-call: 0*/
+                noop.call(undefined);
+            }).not.toThrow();
+            expect(function () {
+                /*eslint no-useless-call: 0*/
```
Comment (ljharb) [inline tests/spec/s-function.js]: nbd, but i find it easier to grep for `/* eslint-disable no-useless-call */` and `/* eslint-enable no-useless-call */`

File: tests/spec/s-function.js [hunk 22]
```diff
@@ -1,7 +1,177 @@
...
+            }).not.toThrow();
+            expect(function () {
+                noop.call(true);
+            }).not.toThrow();
+            expect(function () {
+                noop.call(noop);
+            }).not.toThrow();
+            expect(function () {
+                noop.call(new Date());
+            }).not.toThrow();
```
Comment (ljharb) [inline tests/spec/s-function.js]: i think all these 3-line assertions can be condensed to one line.

File: es5-shim.js (hunk 18 above)
Comment (Xotic750) [inline es5-shim.js]: Ok

File: tests/spec/s-function.js [hunk 23]
```diff
@@ -1,7 +1,177 @@
...
+            expect(function () {
+                noop.call(/test/);
+            }).not.toThrow();
+        });
+        it('should not throw a error with additional args', function () {
```
Comment (ljharb) [inline tests/spec/s-function.js]: what is this protecting against? all functions in JS accept additional arguments - and this isn't testing that `call` is passing the arguments through

File: es5-shim.js (hunk 19 above)
Comment (Xotic750) [inline es5-shim.js]: Ok

File: tests/spec/s-function.js [hunk 24]
```diff
@@ -1,7 +1,177 @@
...
+                noop.call(true, undefined, null, '1', 1, true, noop);
+            }).not.toThrow();
+            expect(function () {
+                noop.call(noop, undefined, null, '1', 1, true, noop);
+            }).not.toThrow();
+            expect(function () {
+                noop.call(new Date(), undefined, null, '1', 1, true, noop);
+            }).not.toThrow();
+        });
+        if (hasStrictMode) {
```
Comment (ljharb) [inline tests/spec/s-function.js]: instead of this if-wrapping, please use the approach I've used in most other test files - something like:
```js
var ifHasStrictIt = hasStrictMode ? it : xit;`
```

(I'll fix the current instance of this pattern in the array tests)

File: tests/spec/s-object.js [hunk 25]
```diff
@@ -356,4 +361,35 @@ describe('Object', function () {
...
+            expect(toStr.call([1, 2, 3])).toBe('[object Array]');
+            expect(toStr.call(returnArgs())).toBe('[object Arguments]');
+            expect(toStr.call({})).toBe('[object Object]');
+            expect(toStr.call(noop)).toBe('[object Function]');
+            expect(toStr.call(new RegExp('c'))).toBe('[object RegExp]');
+            expect(toStr.call(new Date())).toBe('[object Date]');
+            expect(toStr.call(new Error('x'))).toBe('[object Error]');
+        });
+
+        it('prototypes', function () {
```
Comment (ljharb) [inline tests/spec/s-object.js]: this test will actually cause problems in future implementations, since in ES6, many of these prototype objects stop being instances of themselves, and instead would return `'[object Object]'`.

Please split these up such that array/boolean/object/function are always tested, but that the rest are skipped with an `xit`, much like https://github.com/es-shims/es5-shim/blob/master/tests/spec/s-object.js#L89-L90

File: tests/spec/s-object.js [hunk 26]
```diff
@@ -22,6 +22,11 @@ var canFreeze = typeof Object.freeze === 'function' && (function () {
     return obj.a !== 3;
 }());
 var ifCanFreezeIt = canFreeze ? it : xit;
+var toStr = Object.prototype.toString;
+var noop = function () {};
+var returnArgs = function () {
```
Comment (ljharb) [inline tests/spec/s-object.js]: this can probably just be inlined at the one place it's used?

File: tests/spec/s-function.js (hunk 21 above)
Comment (Xotic750) [inline tests/spec/s-function.js]: Aha, I looked at the docs and wasn't sure how to do it, I've never use `eslint` before.

File: tests/spec/s-function.js (hunk 22 above)
Comment (Xotic750) [inline tests/spec/s-function.js]: Ok

File: tests/spec/s-function.js (hunk 23 above)
Comment (Xotic750) [inline tests/spec/s-function.js]: I threw in all the tests that I had. There are sure to be improvements. I'll try and improve them some.

File: tests/spec/s-function.js (hunk 24 above)
Comment (Xotic750) [inline tests/spec/s-function.js]: I'll take a closer look and try to mimic your style.

File: tests/spec/s-object.js (hunk 25 above)
Comment (Xotic750) [inline tests/spec/s-object.js]: I'll take a look and try to do that.

File: tests/spec/s-object.js (hunk 26 above)
Comment (Xotic750) [inline tests/spec/s-object.js]: Ok

File: es5-shim.js (hunk 11 above)
Comment (Xotic750) [inline es5-shim.js]: I'll change it to use `bind`, at present I can't think of another way.

**Comment by ljharb** [comment]
Thanks for bearing with all the comments! Please rebase this on top of latest master as you make your changes as well.

**Comment by Xotic750** [comment]
Will do. Comments are good, I'm learning. ;)

File: es5-shim.js (hunk 19 above)
Comment (Xotic750) [inline es5-shim.js]: Ok, altered some settings in my editor and changed them back, now I'm seeing a warning `this character may get silently deleted by one or more browsers`. Just thought I'd mention it before I go and disable this warning.

Comment (ljharb) [inline es5-shim.js]: Thanks - It's only an issue if text encoding isn't properly set up - I'm not concerned about it at this point.

File: tests/spec/s-function.js (hunk 20 above)
Comment (Xotic750) [inline tests/spec/s-function.js]: Like this?
```
var global = (function () {
    if (this) {
        return this;
    }
}());
```

File: tests/spec/s-function.js [hunk 27]
```diff
@@ -1,7 +1,177 @@
...
+                }).not.toThrow();
+                expect(result).toBe(subject);
+            });
+        }
+        it('should have correct context in non-strict mode', function () {
+            var result,
+                testFn,
+                subject;
+
+            testFn = function () {
```
Comment (ljharb) [inline tests/spec/s-function.js]: we don't use `onevar`, so please use one `var` per var

File: tests/spec/s-function.js [hunk 28]
```diff
@@ -1,7 +1,177 @@
...
+                return this;
+            };
+
+            expect(function () {
+                result = testFn.call();
+            }).not.toThrow();
+
+            expect(result).toBe(global);
```
Comment (ljharb) [inline tests/spec/s-function.js]: in these tests, why not just `expect(testFn.call()).toBe(global)`? in what scenario would `call` ever throw on a non-throwing function?

File: tests/spec/s-function.js (hunk 20 above)
Comment (ljharb) [inline tests/spec/s-function.js]: I'm thinking more that this needs to be a test - since detecting the global object is very difficult.

Alternatively, you could use `var global = Function('return this')();` which is unassailably correct everywhere.

Comment (Xotic750) [inline tests/spec/s-function.js]: Ok, I'll use your suggestion but I have a feeling I read somewhere that this can call problems in some versions of Chrome, but I'm not sure.

File: tests/spec/s-function.js (hunk 27 above)
Comment (Xotic750) [inline tests/spec/s-function.js]: Ok

File: tests/spec/s-function.js (hunk 28 above)
Comment (Xotic750) [inline tests/spec/s-function.js]: I'll change it. Again, just all my old naive tests, they've never been scrutinised by anyone before.

File: tests/spec/s-function.js (hunk 20 above)
Comment (ljharb) [inline tests/spec/s-function.js]: I used it in the es6-shim for a long time - it can cause problems in Chrome OS plugins, because they disallow CSP, but that's not a concern for tests.

File: tests/spec/s-function.js (hunk 28 above)
Comment (ljharb) [inline tests/spec/s-function.js]: in that case the exception would auto-fail the test anyways :-)

**Comment by Xotic750** [comment]
@ljharb Ok, I think I've dealt with all (or at least the majority) of the comments. Feel free to throw some more at me. I still have no idea where I can see the browser tests, so I'm really unable to assess what is going on there but the Node stuff all looks good.

**Comment by Xotic750** [comment]
I managed to get a test to run on IE8 and it failed. I have clearly got something wrong, it doesn't look like the patch is running or something.

File: es5-shim.js [hunk 29]
```diff
@@ -328,6 +351,99 @@ var strSplit = call.bind(StringPrototype.split);
...
+          return '[object Arguments]';
+        }
+        if (hasToStringTagArrayBug && isDuckTypeArray(thisArg)) {
+          return '[object Array]';
+        }
+        if (hasToStringTagRegExpBug && isTryCatchRegExp(thisArg)) {
+          return '[object RegExp]';
+        }
+      }
+      return apply.bind(this, thisArg, baseSlice(arguments, 1))();
```
Comment (ljharb) [inline es5-shim.js]: it seems like we should cache `apply.bind(call)` outside the function, and just invoke that directly here?

Comment (Xotic750) [inline es5-shim.js]: Probably, I will try to make some changes today and see if I can a slot on IE8, perhaps even force the path with a `true` to see what is happening.

**Comment by Xotic750** [comment]
We now have all green on Opera 11 and IE10.
IE8, tests now run but we have recursive problem due to `bind` shim using `call`. Will think about it a bit.

**Comment by Xotic750** [comment]
IE8 `Function#call` now works and we get correct results with `Object#toString`. I unable to test any other old ES3 browser (IE<9, Opera<11 or any that I have not already mentioned). Will rebase now and have a think about `apply`, looks like a challenge.

**Comment by Xotic750** [comment]
I believe I have a solution for `apply`, but I don't know if it would be acceptable? It would mean extending `Function.prototype` with a new method of some `name` that is unlikely to clash with anything a user may do. e.g `__saAdhVjh5wei3C5fibsd34vGs1udf__` or something like that?

**Comment by ljharb** [comment]
Yeah that's definitely not acceptable :-/ Reflection would easily expose it.

File: es5-shim.js [hunk 30]
```diff
@@ -328,6 +329,73 @@ var strSplit = call.bind(StringPrototype.split);
...
 
+// ES-5 15.3.4.4
+// http://es5.github.io/#x15.3.4.4
+var hasToStringTagBasicBug = toStr() !== '[object Undefined]' || toStr(null) !== '[object Null]';
+var hasToStringTagLegacyArguments = toStr(arguments) !== '[object Arguments]';
+var hasToStringTagInconsistency = hasToStringTagBasicBug || hasToStringTagLegacyArguments;
+
+// Others to check:
+// PhantomJS 1.9 which returns 'function' for `NodeList` instances.
+// Older ES3 native functions like `alert`, there are some fixes available but
```
Comment (ljharb) [inline es5-shim.js]: no, i'm not going to worry about older "alert" implementations

File: es5-shim.js [hunk 31]
```diff
@@ -328,6 +329,73 @@ var strSplit = call.bind(StringPrototype.split);
...
+        while (++index < length) {
+            result[index] = array[index + begin];
+        }
+        return result;
+    };
+
+    defineProperties(FunctionPrototype, {
+      call: function () {
+          var thisArg = arguments[0];
+          if (this === to_string) {
```
Comment (ljharb) [inline es5-shim.js]: `to_string` doesn't seem to be defined anywhere

File: es5-shim.js [hunk 32]
```diff
@@ -328,6 +329,73 @@ var strSplit = call.bind(StringPrototype.split);
...
+              if (hasToStringTagBasicBug) {
```
Comment (ljharb) [inline es5-shim.js]: any reason not to unconditionally enter this code branch? If `call` has to be patched, we might as well just do null and undefined for it

File: es5-shim.js (hunk 31 above)
Comment (Xotic750) [inline es5-shim.js]: https://github.com/es-shims/es5-shim/blob/master/es5-shim.js#L63

File: es5-shim.js (hunk 32 above)
Comment (Xotic750) [inline es5-shim.js]: We only want `call` to give us a `[[Class]]` if the context is `Object#toString`, otherwise we want it to go about its normal business of call whatever the function `this` is.

**Comment by Xotic750** [comment]
I'll clean that up, and a variable. I noticed,  that is not used. I'll leave a comment about this is where other native objects can be fixed (even though you don't intend to). I'll also cleanup `baseSlice` into just a loop. I can't see me using it for `apply` for some time, as I can't think of a way to avoid recursion at present without extending `Function#` or attaching (and removing) a property to the individual function so that `apply` can get context.

**Comment by Xotic750** [comment]
The only useful thing remaining, that I can think of, would be for normalising the `[[Class]]` response when checking `window` or `global`. But that is something I can leave you to think(or not) about. ;)

**Comment by Xotic750** [comment]
And I think I've found a solution to the recursion problem and getting context without extending Function# ;)

File: es5-shim.js (hunk 31 above)
Comment (ljharb) [inline es5-shim.js]: ah, my bad, i thought you defined it :-) disregard

**Comment by ljharb** [comment]
Yes, let's not bother with `window` or the global object.

File: es5-shim.js [hunk 33]
```diff
@@ -328,6 +328,54 @@ var strSplit = call.bind(StringPrototype.split);
...
+        !owns(value, 'arguments') &&
+        owns(value, 'callee');
+        // Using `isCallable` here is dangerous as it can cause recursion.
+        // A typeof check would be ok if necessary.
+  };
+}
+
+defineProperties(FunctionPrototype, {
+  call: function () {
+      var thisArg = arguments[0];
```
Comment (ljharb) [inline es5-shim.js]: `Function.call` has a length of `1`, so `thisArg` should be a named argument. http://www.ecma-international.org/ecma-262/5.1/#sec-15.3.4.4

File: es5-shim.js [hunk 34]
```diff
@@ -328,6 +328,54 @@ var strSplit = call.bind(StringPrototype.split);
...
+          }
+          if (typeof thisArg === 'undefined') {
+              return '[object Undefined]';
+          }
+          if (hasToStringTagLegacyArguments && isDuckTypeArguments(thisArg)) {
+              return '[object Arguments]';
+          }
+      }
+      var i = arguments.length;
+      var rest = i ? Array(--i) : [];
```
Comment (ljharb) [inline es5-shim.js]: please explicitly compare to `0` instead of relying on truthiness - however, i'm not sure why we need `Array()` at all? There's no performance benefit towards "allocating" slots on a JS array, in fact the opposite - because it initially creates a sparse array.

File: es5-shim.js (hunk 33 above)
Comment (Xotic750) [inline es5-shim.js]: Ok

File: es5-shim.js (hunk 34 above)
Comment (Xotic750) [inline es5-shim.js]: I though it was cheaper than using `push`, if you have a large number or arguments?

File: es5-shim.js [hunk 35]
```diff
@@ -56,9 +56,17 @@ var array_push = ArrayPrototype.push;
 var array_unshift = ArrayPrototype.unshift;
 var array_concat = ArrayPrototype.concat;
 var call = FunctionPrototype.call;
+var apply = FunctionPrototype.apply;
 var max = Math.max;
 var min = Math.min;
 
+// To prevent recursion if `Function#call` or `Function#apply` are patched.
```
Comment (Xotic750) [inline es5-shim.js]: This actually needs to be moved to happen on when patch is applied. I will fix it an rebase later.

File: es5-shim.js [hunk 36]
```diff
@@ -328,6 +329,70 @@ var strSplit = call.bind(StringPrototype.split);
...
+// Tests for inconsistent or buggy `[[Class]]` strings.
+var hasToStringTagBasicBug = toStr() !== '[object Undefined]' || toStr(null) !== '[object Null]';
+var hasToStringTagLegacyArguments = toStr(arguments) !== '[object Arguments]';
+var hasToStringTagInconsistency = hasToStringTagBasicBug || hasToStringTagLegacyArguments;
+// Others that could be fixed:
+// Older ES3 native functions like `alert` return `[object Object]`.
+// Inconsistent `[[Class]]` strings for `window` or `global`.
+
+if (hasToStringTagLegacyArguments) {
+    // To prevent recursion if `Function#call` or `Function#apply` are patched.
```
Comment (Xotic750) [inline es5-shim.js]: This actually seems to have some strange effect on newer environments (when patch is forced) when not used in conjunction with my `apply` patch. I'm thinking that I may remove it from this this PR.

**Comment by Xotic750** [comment]
I am going to remove the `Array#slice` test from this PR and submit those separately.

**Comment by Xotic750** [comment]
I now have apply working on IE8
https://rawgit.com/Xotic750/es5-shim/apply/tests/index.html
Ref: https://github.com/es-shims/es5-shim/issues/99
I feel it is more robust to implement both fixes together. What are your thoughts on how to proceed?

**Comment by Xotic750** [comment]
`call` + `apply` fix is far more stable.
